pennylane_lightning/src/util/Util.hpp

```diff
@@ -915,6 +915,80 @@ auto linspace(T start, T end, size_t num_points) -> std::vector<T> {
     return data;
 }
 
+/**
+ * @brief Determines the indices that would sort an array.
+ *
+ * @tparam T vector data type.
+ * @param arr Array to be inspected.
+ * @return a vector with indices that would sort the array.
+ */
+template <typename T>
+inline auto sorting_indexes(const T &arr, size_t length)
+                            -> std::vector<size_t>  {
+  std::vector<size_t> indexes(length);
+  iota(indexes.begin(), indexes.end(), 0);
+
+  // indexes will be sorted in accordance to the array provided.
+  sort(indexes.begin(), indexes.end(),
+       [&arr](size_t i1, size_t i2) {return arr[i1] < arr[i2];});
+
+  return indexes;
+}
+
+/**
+ * @brief Determines the indices that would sort a vector.
+ *
+ * @tparam T array data type.
+ * @param vec Vector to be inspected.
+ * @return a vector with indices that would sort the vector.
+ */
+template <typename T>
+inline auto sorting_indexes(const std::vector<T> &vec) 
+                            -> std::vector<size_t> {
+  return sorting_indexes(vec.data(), vec.size());
+}
+
+/**
+ * @brief Determines a new index for a transposed tensor stored linearly.
+ *  This function assumes each axis will have a length of 2 (|0>, |1>).
+ *
+ * @param ind index before.
+ * @param new_axes new axes distribution.
+ * @return unsigned int with the new transposed index.
+ */
+inline auto transposed_state_index (  size_t ind, 
+                                        const std::vector<size_t> &new_axes)
+                                    -> size_t {
+    size_t new_index=0;
+    for (std::vector<size_t>::const_reverse_iterator axis = new_axes.rbegin(); 
+            axis != new_axes.rend(); ++axis){
+        new_index+=(ind%2)*pow(2,(new_axes.size()-1)-(*axis));
+        ind/=2;
+    }
+    return new_index;
+}
+
+/**
+ * @brief Template for the transposition of state tensors, 
+ * axes are assumed to have a length of 2 (|0>, |1>).
+ *
+ * @tparam T tensor data type.
+ * @param tensor Tensor to be transposed.
+ * @param new_axes new axes distribution.
+ * @return Transposed Tensor.
+ */
+template <typename T>
+auto transpose_state_tensor( 
+                            const std::vector<T> &tensor, 
+                            const std::vector<size_t> &new_axes)
+                            -> std::vector<T> {
+    std::vector<T> transposed_tensor(tensor.size());
+    for (size_t ind=0; ind<tensor.size(); ind++){
+        transposed_tensor[ind]=tensor[transposed_state_index(ind, new_axes)];
+    }
+  return transposed_tensor;
+}
```

Comment:
I worry this may be slow for large data sets. Can you run some numbers on this and see how it performs for various numbers of qubits.

Reply:
Yes. Sure.

Reply:
I explored some alternative strategies for this transposition, and at least for now, this seems to be the more efficient. In the other hand, It doesn't seem to be too slow.
I checked the performance for 100 runs, up to 22 qubits.
I applied the `transpose_state_tensor` function to vectors with the size of the equivalent state vector (or tensor). This shows exactly the information we are interested in. For 22 qubits, we have a vector with 8,388,608 elements, and the mean time is 0.89 seconds.

![image](https://user-images.githubusercontent.com/87949283/145645265-2d3d04de-6e7d-4092-854f-6fdffa902835.png)

A loop like that, with no data dependency, is quite the low-hanging fruit for openMP parallelisation, but for the sizes I'm presenting here quite of an overkill. If you think we should proceed with some parallelisation strategy, I think we could impose a threshold based in the number of qubits.

Reply:
Thanks Amintor. I do think we should ensure this is fast  for <=30 qubits (for large node usage). We can add parallelism as a next pass though.

Reply:
Hi @mlxd ,
I extended the benchmark to 26 qubits (memory is limiting this range).
I optimised the code to determine the transposed index, and the function to transpose the tensor, accordingly.
The best method so far is still based in transpose the tensor after the calculation of the probabilities.
Anyway, in this new approach for the tensor transposition, I could observe a huge increase in speed, for larger number of qubits. I'm comparing the two methods here:
![download](https://user-images.githubusercontent.com/87949283/145894671-ac019440-9447-4a8f-8e2a-10b87bf8258b.png)
I also plotted the ration between the two times:
![download](https://user-images.githubusercontent.com/87949283/145894723-21541e59-2219-48e0-9525-1950db52220b.png)

For 26 qubits, the previous algorithm is taking 18.4 seconds, while the new one is taking 2.3 seconds. The new code is approximately 8 times faster. This qubit state vector has 67,108,864 elements. I'm performing calculation with double precision, and compiling with the -Ofast flag.

We can observe an outlier in 12 qubits, where the previous algorithm seems to outperform the new one. The ration here is 0.85, and this may be because of the number of samples (100). Also, we are talking about small number and this could be strongly affected by fluctuations.

Reply:
In other run, the 12 qubits performed better in the new algorithm. I think it is reasonable to assume that the measure of times in this fast cases is specially prone to fluctuations. It seems that starting in 20 qubits would be a good choice.